src/actions/spring.js

```diff
@@ -0,0 +1,93 @@
+/*
+  The closed-form damped harmonic oscillating spring.
+  Or, spring.
+
+  This is a direct port of Adam Miskiewicz's (@skevy) React Animated
+  PR #15322 https://github.com/facebook/react-native/pull/15322/
+
+  ```
+  spring({
+    mass: 2,
+    damping: 10,
+    stiffness: 100,
+    to: 100
+  }).start();
+  ```
+
+  Adam Miskiewicz:
+  @skevy (twitter.com/skevy, github.com/skevy)
+ */
+import Action from './';
+import { timeSinceLastFrame } from '../framesync';
+
+class Spring extends Action {
+  static defaultProps = {
+    stiffness: 100,
+    damping: 10,
+    mass: 1.0,
+    velocity: 0.0,
+    from: 0.0,
+    to: 0.0,
+    restSpeed: 0.01,
+    restDisplacement: 0.01
+  };
+
+  onStart() {
+    const { velocity, to } = this.props;
+    this.t = 0;
+    this.initialVelocity = velocity / 1000;
+    this.isComplete = false;
+    this.delta = to - this.from;
+  }
+
+  update() {
+    const { stiffness, damping, mass, from, to, restSpeed, restDisplacement } = this.props;
+    const { delta, initialVelocity } = this;
+
+    const timeDelta = timeSinceLastFrame() / 1000;
+    const t = this.t = this.t + timeDelta;
+
+    const dampingRatio = damping / (2 * Math.sqrt(stiffness * mass));
+    const angularFreq = Math.sqrt(stiffness / mass);
+    const expoDecay = angularFreq * Math.sqrt(Math.abs(1.0 - (dampingRatio * dampingRatio)));
```

Comment:
The error is in your "critically damped" branch (I pointed it out below)...`expoDecay` is only used when the spring is underdamped (in which case the math will work out correctly, because dampingRatio^2 will be less than one).